Intermittent build error - CS0016: Could not write to output file (path not found)

### Description

Sometimes during our builds, we encounter the following error:

```
CSC : error CS0016: Could not write to output file 'D:\a\_work\1\s\artifacts\obj\Core\Release\net10.0-windows10.0.20348.0\generated\System.Text.Json.SourceGeneration\System.Text.Json.SourceGeneration.JsonSourceGenerator\HybridWebViewHandlerJsonContext.GetJsonTypeInfo.g.cs' -- 'Could not find a part of the path 'D:\a\_work\1\s\artifacts\obj\Core\Release\net10.0-windows10.0.20348.0\generated\System.Text.Json.SourceGeneration\System.Text.Json.SourceGeneration.JsonSourceGenerator\HybridWebViewHandlerJsonContext.GetJsonTypeInfo.g.cs'.' [D:\a\_work\1\s\src\Core\src\Core.csproj::TargetFramework=net10.0-windows10.0.20348.0]
```

This happens sporadically in CI builds. A typical failed run can be seen here:

[Link to failed build](https://dev.azure.com/dnceng/internal/internal%20Team/_build/results?buildId=2836272&view=logs&j=7d201231-4672-5a26-415a-d34762af3de0&t=5f4dfcc8-f694-5357-5d6b-12cb82f51270&l=4424)

It appears to be a filesystem or IO issue during code generation.

### Steps to Reproduce

1. Trigger a CI build.
2. Observe that, sporadically, the build breaks with CS0016 errors on file write in the Core project.
3. See the [linked build log](https://dev.azure.com/dnceng/internal/internal%20Team/_build/results?buildId=2836272&view=logs&j=7d201231-4672-5a26-415a-d34762af3de0&t=5f4dfcc8-f694-5357-5d6b-12cb82f51270&l=4424) for an occurrence.

Expected: Build completes successfully.
Actual: Build fails with CS0016 path/file error intermittently.

### Link to public reproduction project repository

https://dev.azure.com/dnceng/internal/internal%20Team/_build/results?buildId=2836272&view=logs&j=7d201231-4672-5a26-415a-d34762af3de0&t=5f4dfcc8-f694-5357-5d6b-12cb82f51270&l=4424

### Version with bug

10.0.0

### Is this a regression from previous behavior?

Not sure, did not test other versions

### Last version that worked well

Unknown/Other

### Affected platforms

Windows

### Affected platform versions

Windows Server 2022 (Azure pipeline agent)

### Did you find any workaround?

Sometimes restarting the pipeline agent or clearing the build folders before build helps, but the issue can still occur unpredictably.

### Relevant log output

```shell
CSC : error CS0016: Could not write to output file 'D:\a\_work\1\s\artifacts\obj\Core\Release\net10.0-windows10.0.20348.0\generated\System.Text.Json.SourceGeneration\System.Text.Json.SourceGeneration.JsonSourceGenerator\HybridWebViewHandlerJsonContext.GetJsonTypeInfo.g.cs' -- 'Could not find a part of the path 'D:\a\_work\1\s\artifacts\obj\Core\Release\net10.0-windows10.0.20348.0\generated\System.Text.Json.SourceGeneration\System.Text.Json.SourceGeneration.JsonSourceGenerator\HybridWebViewHandlerJsonContext.GetJsonTypeInfo.g.cs'.' [D:\a\_work\1\s\src\Core\src\Core.csproj::TargetFramework=net10.0-windows10.0.20348.0]
```

--- a/Directory.Build.props
+++ b/Directory.Build.props
@@ -277,4 +277,9 @@
     <!-- Disables the transitive restore of packages like Microsoft.AspNetCore.App.Ref, Microsoft.WindowsDesktop.App.Ref -->
     <DisableTransitiveFrameworkReferenceDownloads>true</DisableTransitiveFrameworkReferenceDownloads>
   </PropertyGroup>
+
+   <!-- Workaround for EmitCompilerGeneratedFiles issue https://microsoft.visualstudio.com/OS/_workitems/edit/57204991 -->
+  <PropertyGroup>
+    <CoreCompileDependsOn>$(CoreCompileDependsOn);CreateCompilerGeneratedFilesOutputPath</CoreCompileDependsOn>
+  </PropertyGroup>
 </Project>

